app/services/courses/publish_service.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module Courses
+  class PublishService
+    UnpublishableError = Class.new(StandardError)
+
+    class << self
+      def publish(uuid:, user:)
+        Rails.logger.tagged('Course::PublishService') do |l|
+          l.info "publishing course uuid: #{uuid}"
+        end
+        course = Course.find_by(uuid:)
+
+        raise UnpublishableError unless course.publishable?
```

Comment:
I don't think it should raise an error. I want to use this in the controller so I don't want to catch this error.

Should this
1) return false if the course is unpulishable?
2) return the course and expect the caller to check if the course was published?

Reply:
I'd go for the solution 2.

This is how I'd write this:

```ruby
module Courses
  class CoursePublisher
    def initialize(course:, user:)
      @course = course
      @user = user
    end

    def call
      return false unless @course.publishable?

      publish_course
      notify_course_published

      @course
    end

    private

    def publish_course
      Course.transaction do
        @course.publish_sites
        @course.publish_enrichment(@user)
        @course.application_status_open!
      end
    end

    def notify_course_published
      NotificationService::CoursePublished.call(course: @course)
    end
  end
end
```

And on controller:

```ruby
def publish
  course = Course.find_by(uuid: params[:uuid])
  
  if result = Courses::CoursePublisher.new(course: course, user: current_user).call
    redirect_to course_path(result), notice: 'Course published successfully!'
  else
    redirect_to courses_path, alert: 'Course could not be published.'
  end
end
```

I put the notification out of the transaction. Not sure if it should though. 

Also Publish has this pattern of `include ServicePattern`. Not sure if I like to suffix everything with `Service` though.

Reply:
Nice, I'll incorporate this into the PR